UnityProject/Assets/Scripts/Items/Tool/TRayScanner.cs

```diff
@@ -0,0 +1,214 @@
+		}
+
+		public void ServerPerformInteraction(HandActivate interaction)
+		{
+			IncrementMode();
+
+			if (currentMode == Mode.Off)
+			{
+				Chat.AddExamineMsgFromServer(interaction.Performer, $"You switch the {gameObject.ExpensiveName()} off");
+				spriteHandler.ChangeSprite(0);
+				return;
+			}
+
+			Chat.AddExamineMsgFromServer(interaction.Performer,
+				$"You switch the {gameObject.ExpensiveName()} to detect {currentMode.ToString()}");
+			spriteHandler.ChangeSprite(1);
+		}
+
+		private void IncrementMode()
+		{
+			var currentModeInt = (int)currentMode;
+			currentModeInt++;
+
+			if (currentModeInt > (types.Length - 1))
+			{
+				currentModeInt = 0;
+			}
+
+			currentMode = (Mode)currentModeInt;
+		}
+
+		public void OnInventoryMoveServer(InventoryMove info)
+		{
+			//From a player
+			if (info.FromPlayer != null)
+			{
+				//Removed or put in backpack
+				if (info.ToSlot == null || info.ToSlot.Player == null)
+				{
+					//Turn off
+					DoRpc(info.FromPlayer, false);
+				}
+			}
+
+			if(info.FromPlayer == info.ToPlayer) return;
+
+			//Put on ground dont need to go further
+			if(info.ToSlot == null) return;
+
+			//To new player
+			if (info.ToPlayer != null)
+			{
+				//Put in backpack
+				if (info.ToSlot.Player == null)
+				{
+					//Should already be off so dont need to do anything
+					return;
+				}
+
+				//Ping to turn on if needed
+				DoRpc(info.ToPlayer, currentMode != Mode.Off);
+			}
+		}
+
+		private void DoRpc(RegisterPlayer player, bool newState)
+		{
+			if(player.connectionToClient == null) return;
+
+			if(CustomNetworkManager.IsServer && CustomNetworkManager.IsHeadless == false)
+			{
+				//Target RPC not working on local host?
+				if (newState)
+				{
+					DoState(currentMode);
+					return;
+				}
+
+				DoState(Mode.Off);
+				return;
+			}
+
+			RpcChangeState(player.connectionToClient, newState);
+		}
+
+		[TargetRpc]
+		private void RpcChangeState(NetworkConnection conn, bool newState)
+		{
+			if (newState)
+			{
+				DoState(currentMode);
+				return;
+			}
+
+			DoState(Mode.Off);
+		}
+
+		private void SyncMode(Mode oldMode, Mode newMode)
+		{
+			currentMode = newMode;
+			if(PlayerManager.LocalPlayerScript == null) return;
+
+			if(pickupable.ItemSlot == null || pickupable.ItemSlot.Player == null ||
+			   pickupable.ItemSlot.Player != PlayerManager.LocalPlayerScript.registerTile) return;
+
+			DoState(currentMode);
+		}
+
+		private void DoState(Mode newMode)
+		{
+			var matrixInfos = MatrixManager.Instance.ActiveMatricesList;
+
+			//TODO update for layer PR
+
+			foreach (var matrixInfo in matrixInfos)
+			{
+				var tilemapRenderer = matrixInfo.Matrix.UnderFloorLayer.GetComponent<TilemapRenderer>();
+				tilemapRenderer.sortingLayerName = newMode == Mode.Off ? "UnderFloor" : "Walls";
+				tilemapRenderer.sortingOrder = newMode == Mode.Off ? 0 : 1;
+			}
+		}
+
+		public void OnPlayerRejoin()
+		{
+			if(pickupable.ItemSlot?.Player == null) return;
```

Comment:
we need a better system for something like this , that handles removing when taken out a player's inventory, and reloging and all that jizz